src/Lecture2.hs

```diff
@@ -215,8 +326,13 @@ The algorithm of merge sort is the following:
 [1,2,3]
 -}
 mergeSort :: [Int] -> [Int]
-mergeSort = error "TODO"
-
+mergeSort [] = []
+mergeSort [x] = [x]
+mergeSort xs =
+  let (l, r) = splitAt (length xs `div` 2) xs
```

Comment:
You can avoid using `length` if you notice that you don't necessarily need to split in the middle. You only need to split it into two lists of approximately the same size.

In that case, you can implement a solution that traverses a list only once and splits its elements into two lists of elements on only even positions and only odd positions.

Alternatively, you can implement a bottom-up approach by converting a list of elements to a list of of singleton lists and merging them pairwise until only one list is left.

Reply:
Ah I see because both length and splitAt have to traverse the list. I tried implementing a function that divides the list in halve by traversing it just once with linear complexity:
```haskell
splitHalves :: [a] -> ([a], [a])
splitHalves = go [] []
 where
  go h1 h2 [] = (h1, h2)
  go h1 h2 (x : xs) = go h2 (x : h1) xs
```

Also I noticed the recursive step of the `mergeSort` could be rewritten as an elegant one liner :)

```haskell
mergeSort [] = []
mergeSort [x] = [x]
mergeSort xs = uncurry merge $ both mergeSort $ splitHalves xs
  where both f (a1, a2) = (f a1, f a2)
```